pkg/hive/hive.go

```diff
@@ -0,0 +1,297 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package hive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path"
+	"reflect"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"go.uber.org/fx"
+	"go.uber.org/fx/fxevent"
+	"go.uber.org/fx/fxtest"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "hive")
+)
+
+const (
+	// defaultStartTimeout is the amount of time allotted for start hooks. After
+	// this duration the context passed to the start hooks is cancelled.
+	defaultStartTimeout = 5 * time.Minute
+
+	// defaultStopTimeout is the amount of time allotted for stop hooks.
+	defaultStopTimeout = time.Minute
+
+	// defaultEnvPrefix is the default prefix for environment variables, e.g.
+	// flag "foo" can be set with environment variable "CILIUM_FOO".
+	defaultEnvPrefix = "CILIUM_"
+)
+
+// Hive is a modular application built from cells.
+type Hive struct {
+	app                       *fx.App
+	cells                     []*Cell
+	dotGraph                  fx.DotGraph
+	envPrefix                 string
+	flags                     *pflag.FlagSet
+	fxLogger                  *logging.FxLogger
+	lifecycle                 lifecycleProxy
+	startTimeout, stopTimeout time.Duration
+	viper                     *viper.Viper
+}
+
+// New returns a new hive that can be run, or inspected.
+// The command-line flags from the cells are registered as part of this.
+//
+// The object graph is not constructed until methods of the hive are
+// invoked.
+func New(v *viper.Viper, flags *pflag.FlagSet, cells ...*Cell) *Hive {
+	h := &Hive{
+		envPrefix:    defaultEnvPrefix,
+		fxLogger:     logging.NewFxLogger(log),
+		cells:        cells,
+		viper:        v,
+		flags:        pflag.NewFlagSet("", pflag.ContinueOnError),
+		startTimeout: defaultStartTimeout,
+		stopTimeout:  defaultStopTimeout,
+	}
+	if err := h.registerFlags(flags); err != nil {
+		log.Fatal(err)
+	}
+	return h
+}
+
+func (h *Hive) SetTimeouts(start, stop time.Duration) {
+	h.startTimeout, h.stopTimeout = start, stop
+}
+
+func (h *Hive) SetEnvPrefix(prefix string) {
+	h.envPrefix = prefix
+}
+
+// Run populates the cell configurations and runs the hive cells.
+// If an error occurs when populating cell configurations this
+// method panics.
+// Interrupt signal will cause the hive to stop.
+func (h *Hive) Run() {
+	if h.app == nil {
+		if err := h.createApp(); err != nil {
+			log.Fatalf("populate: %s", err)
```

Comment:
nit: error may come from `h.app().Err()` too, so we should probably log something more general like "create app: ..." instead of "populate: ..."

Reply:
Fixed.